Zend/zend_vm_def.h

```diff
@@ -2085,7 +2086,15 @@ ZEND_VM_C_LABEL(fetch_obj_r_fast_copy):
 							ZEND_VM_NEXT_OPCODE();
 						}
 					}
+				} else if (UNEXPECTED(IS_HOOKED_PROPERTY_OFFSET(prop_offset))) {
+					if (ZEND_IS_PROPERTY_HOOK_SIMPLE_READ(prop_offset)) {
```

Comment:
I might be missing something, but why do we cache "simple read offset" instead of simply caching the actual offset of the backing property?
Like, we do always anyway go to the backing property with these fetches, don't we?

In zend_read_property_ex set `cache_slot[1]` to `prop_info->offset` directly instead of refetching in VM on every access?

Reply:
I had the same thought, and a possible reason for this design is that read and write OPs may share the same cache slot after compaction, so updating `cache_slot[1]` in `zend_read_property_ex` would also affect write operations. If this is the only reason we could avoid compacting reads and writes to the same slot, or have two separate offsets in the cache slot (e.g. cache_slot[1] for reads and cache_slot[2] for writes), but this may negatively impact all property accesses.

Reply:
You're indeed right, however this compaction only happens for op1_type == IS_UNUSED, i.e. for $this fetches. At that point, can't we just take property hooks into account on compaction, as we're aware of the CE?
The only time we cannot do that is for Closures and traits. Possibly fine to just have dedicated r&w slots then.

Reply:
We can't reliably know about hooks of a given property, given that it may be defined in a parent class. But we could separate the cache slot based on the fetch type of the OP itself.

Reply:
For parent ones yes, as said - they would be part of "unknown properties". Would just apply to properties of the class itself - which still will cover the vast majority of all $this-> based fectches in code I'd assume.

Reply:
This check is only reliable on final properties though (or final class), which will be much more rare. A property fetch can be influenced from a child class, when adding hooks to existing properties.

Reply:
I had another look. Apart from the literal compaction issue, some places like `zend_pre_incdec_overloaded_property()` share the same cache slot for `BP_VAR_R` and `BP_VAR_W` operations. Caching the offset for the former (in case we don't have a `get` hook) will re-use the cache for the latter, skipping the `set` hook. The same applies to the `_OP` handlers, and possibly others. These cases would require separating the cache so that one operation can't influence the other. For now, I think it's better to keep the current approach, as it's not too complicated and is less likely to break existing code.